[2024-05-22T00:48:54.038+0000] {taskinstance.py:1159} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: fraud_bitcoin.train_knn scheduled__2024-05-21T00:00:00+00:00 [queued]>
[2024-05-22T00:48:54.076+0000] {taskinstance.py:1159} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: fraud_bitcoin.train_knn scheduled__2024-05-21T00:00:00+00:00 [queued]>
[2024-05-22T00:48:54.080+0000] {taskinstance.py:1361} INFO - Starting attempt 6 of 7
[2024-05-22T00:48:54.137+0000] {taskinstance.py:1382} INFO - Executing <Task(PythonOperator): train_knn> on 2024-05-21 00:00:00+00:00
[2024-05-22T00:48:54.299+0000] {standard_task_runner.py:57} INFO - Started process 4540 to run task
[2024-05-22T00:48:54.311+0000] {standard_task_runner.py:84} INFO - Running: ['***', 'tasks', 'run', 'fraud_bitcoin', 'train_knn', 'scheduled__2024-05-21T00:00:00+00:00', '--job-id', '402', '--raw', '--subdir', 'DAGS_FOLDER/fraud.py', '--cfg-path', '/tmp/tmp952twlyd']
[2024-05-22T00:48:54.336+0000] {standard_task_runner.py:85} INFO - Job 402: Subtask train_knn
[2024-05-22T00:48:54.717+0000] {task_command.py:416} INFO - Running <TaskInstance: fraud_bitcoin.train_knn scheduled__2024-05-21T00:00:00+00:00 [running]> on host 66abfe055500
[2024-05-22T00:48:55.303+0000] {taskinstance.py:1662} INFO - Exporting env vars: AIRFLOW_CTX_DAG_OWNER='***' AIRFLOW_CTX_DAG_ID='fraud_bitcoin' AIRFLOW_CTX_TASK_ID='train_knn' AIRFLOW_CTX_EXECUTION_DATE='2024-05-21T00:00:00+00:00' AIRFLOW_CTX_TRY_NUMBER='6' AIRFLOW_CTX_DAG_RUN_ID='scheduled__2024-05-21T00:00:00+00:00'
[2024-05-22T00:48:55.772+0000] {xcom.py:661} ERROR - Object of type KNeighborsClassifier is not JSON serializable. If you are using pickle instead of JSON for XCom, then you need to enable pickle support for XCom in your *** config or make sure to decorate your object with attr.
[2024-05-22T00:48:55.778+0000] {taskinstance.py:1937} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/json.py", line 91, in default
    return serialize(o)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/serialization/serde.py", line 178, in serialize
    raise TypeError(f"cannot serialize object of type {cls}")
TypeError: cannot serialize object of type <class 'sklearn.neighbors._classification.KNeighborsClassifier'>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/operators/python.py", line 192, in execute
    return_value = self.execute_callable()
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/operators/python.py", line 209, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/opt/airflow/dags/fraud.py", line 109, in train_knn
    ti.xcom_push(key='knn_model', value=knn_model)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/session.py", line 79, in wrapper
    return func(*args, session=session, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 2479, in xcom_push
    XCom.set(
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/session.py", line 76, in wrapper
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/xcom.py", line 244, in set
    value = cls.serialize_value(
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/xcom.py", line 659, in serialize_value
    return json.dumps(value, cls=XComEncoder).encode("UTF-8")
  File "/usr/local/lib/python3.8/json/__init__.py", line 234, in dumps
    return cls(
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/json.py", line 104, in encode
    return super().encode(o)
  File "/usr/local/lib/python3.8/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/lib/python3.8/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/json.py", line 93, in default
    return super().default(o)
  File "/usr/local/lib/python3.8/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type KNeighborsClassifier is not JSON serializable
[2024-05-22T00:48:55.811+0000] {taskinstance.py:1400} INFO - Marking task as UP_FOR_RETRY. dag_id=fraud_bitcoin, task_id=train_knn, execution_date=20240521T000000, start_date=20240522T004854, end_date=20240522T004855
[2024-05-22T00:48:55.844+0000] {standard_task_runner.py:104} ERROR - Failed to execute job 402 for task train_knn (Object of type KNeighborsClassifier is not JSON serializable; 4540)
[2024-05-22T00:48:55.919+0000] {local_task_job_runner.py:228} INFO - Task exited with return code 1
[2024-05-22T00:48:56.029+0000] {taskinstance.py:2778} INFO - 0 downstream tasks scheduled from follow-on schedule check
